rules_gapic/bazel/src/main/java/com/google/api/codegen/bazel/ApisVisitor.java

```diff
@@ -0,0 +1,172 @@
+
+    String packageLocation = file.getParent().toString();
+    ApiVersionedDir bp = bazelApiVerPackages.get(packageLocation);
+    ApiDir bap = bazelApiPackages.get(packageLocation);
+
+    if (bp == null || bap == null) {
+      return FileVisitResult.CONTINUE;
+    }
+
+    String fileName = file.getFileName().toString();
+    if (fileName.endsWith(".yaml")) {
+      if (!fileName.endsWith(".legacy.yaml")) {
+        String fileBody = readFile(file);
+        bp.parseYamlFile(fileName, fileBody);
+        bap.parseYamlFile(fileName, fileBody);
+      }
+    } else if (fileName.endsWith(".proto")) {
+      bp.parseProtoFile(fileName, readFile(file));
+    } else if (fileName.endsWith(".bazel")) {
+      // merge files?
+    } else if (fileName.endsWith(".json")) {
+      bp.parseJsonFile(fileName, readFile(file));
+    }
+
+    return FileVisitResult.CONTINUE;
+  }
+
+  @Override
+  public FileVisitResult postVisitDirectory(Path dir, IOException e) throws IOException {
+    if (!writer) {
+      return FileVisitResult.CONTINUE;
+    }
+
+    String dirStr = dir.toString();
+    ApiVersionedDir bp = bazelApiVerPackages.get(dirStr);
+    BazelBuildFileTemplate template = null;
+    String tmplType = "";
+    if (bp.getProtoPackage() != null) {
+      if (bp.getGapicYamlPath() != null) {
+        bp.injectFieldsFromTopLevel();
+        template = this.gapicApiTempl;
+        tmplType = "GAPIC_VERSIONED";
+      } else if (!bp.getLangProtoPackages().isEmpty()) {
+        template = this.rawApiTempl;
+        tmplType = "RAW";
+      }
+    } else if (bp.getServiceYamlPath() != null) {
+      template = this.rootApiTempl;
+      tmplType = "API_ROOT";
+    }
+
+    if (template == null) {
+      return FileVisitResult.CONTINUE;
+    }
+
+    String rootDirStr = srcDir.toString();
+    String outDirPath = destDir.toString() + dirStr.substring(rootDirStr.length());
+    File outDir = new File(outDirPath);
+
+    if (!outDir.exists()) {
+      if (!outDir.mkdirs()) {
+        System.out.println("WARNING: Could not create directory: " + outDir.toString());
+        return FileVisitResult.CONTINUE;
+      }
+    }
+
+    // System.out.println(bp);
+    System.out.println(
+        "Write File [" + tmplType + "]: " + outDir.toString() + File.separator + "BUILD.bazel");
+    try {
+      BazelBuildFileView bpv = new BazelBuildFileView(bp);
+      fileWriter.write(Paths.get(outDir.toString(), "BUILD.bazel"), template.expand(bpv));
+    } catch (RuntimeException ex) {
+      ex.printStackTrace();
+    }
+
+    return FileVisitResult.CONTINUE;
+  }
+
+  public static String readFile(String path) throws IOException {
+    return new String(Files.readAllBytes(Paths.get(path).normalize()), StandardCharsets.UTF_8);
+  }
+
+  private static String readFile(Path path) throws IOException {
+    return new String(Files.readAllBytes(path), StandardCharsets.UTF_8);
+  }
+
+  public void setWriter(boolean writer) {
+    this.writer = writer;
```

Comment:
Renamed it to `writerMode`. This field determines if the AisVisitor is in writer or reader mode. The way it works: it first walks the file tree to read stuff and then walks same file tree to write stuff. 

The reason why two passes are needed is because the the order in which siblings of a directory are visited is not defined, and some of the properties depend on the "full picture" (i.e. it needs to know all files in dir to pick the right one, like if there are service.yaml and service_v1.yaml for v1 dir it should pick v1). Only doing it in two passes we can guarantee that the "writer" has "full picture" of the previous read directory and pick the correct files and propeerties.